### PR TITLE
Symbols

### LINKED ISSUE #110: SVG symbols not displaying in PDF
I'm using jspdf-yworks and svg2pdf.js to create a PDF. Here is a jsfiddle with an example: https://jsfiddle.net/paulnordlund/9k50jtrx/8/

I'm expecting labels (e.g. title "Phosphosite Intensity") to display on the downloaded PDF, but they are not.  Is this a bug? If not, could you help me understand why symbols instantiated by use elements (<use y="14.732422" x="30.789062" xlink:href="#3951_1-0-glyph3-1>) are not rendered on the PDF?

### PR DESCRIPTION
**Pull request by zenker981**
Implement symbols an update <use>-handling; closes #110

### CODE REVIEW

File: src/svg2pdf.js [hunk 1]
```diff
@@ -804,16 +837,7 @@ SOFTWARE.
         pf(getAttribute(node, "height")) || (vb && vb[3]) || 0
       ];
     } else if (nodeIs(node, "g,clippath")) {
-      boundingBox = [0, 0, 0, 0];
```
Comment (HackbrettXXX) [inline src/svg2pdf.js]: var keyword ;)

File: src/svg2pdf.js [hunk 2]
```diff
@@ -279,6 +279,12 @@ SOFTWARE.
       _pdf.beginFormObject(bBox[0], bBox[1], bBox[2], bBox[3], tfMatrix);
       renderChildren(node, _pdf.unitMatrix, this, false, false, AttributeState.default());
       _pdf.endFormObject(node.getAttribute("id"));
+    } else if (nodeIs(node, "symbol,svg")) {
+      // create unclipped and untransformed form object
+      var bBox = getUntransformedBBox(node);
```
Comment (HackbrettXXX) [inline src/svg2pdf.js]: we should probably use a larger bBox here. At least in the svg case, this is the box determined by width/height or the viewBox.

File: src/svg2pdf.js [hunk 3]
```diff
@@ -531,6 +537,22 @@ SOFTWARE.
       } else {
         nodeTransform = new _pdf.Matrix(1, 0, 0, 1, -x, -y);
       }
+    } else if (nodeIs(node, "symbol")) {
+      x = parseFloat(getAttribute(node, "x")) || 0;
+      y = parseFloat(getAttribute(node, "y")) || 0;
+      // TODO:
+      // x += parseFloat(node.getAttribute("refX")) || 0; ???
```
Comment (HackbrettXXX) [inline src/svg2pdf.js]: what about this? Would be nice if this would work, too

File: src/svg2pdf.js [hunk 4]
```diff
@@ -846,7 +871,7 @@ SOFTWARE.
       ];
     }
 
-    if (!nodeIs(node, "marker,svg,g")) {
+    if (!nodeIs(node, "marker,svg,g")) {  //  TODO: and !symbol ???
```
Comment (HackbrettXXX) [inline src/svg2pdf.js]: add symbol to the list ;)

File: src/svg2pdf.js [hunk 5]
```diff
@@ -863,6 +888,20 @@ SOFTWARE.
     return boundingBox;
   };
 
+  var getBoundingBoxByChildren = function (node) {
+    boundingBox = [0, 0, 0, 0];
```
Comment (HackbrettXXX) [inline src/svg2pdf.js]: var boundingBox = ...

File: src/svg2pdf.js [hunk 6]
```diff
@@ -1209,24 +1248,60 @@ SOFTWARE.
   // draws the element referenced by a use node, makes use of pdf's XObjects/FormObjects so nodes are only written once
   // to the pdf document. This highly reduces the file size and computation time.
   var use = function (node, tfMatrix, refsHandler) {
+    var pf = parseFloat;
```
Comment (HackbrettXXX) [inline src/svg2pdf.js]: there are a lot of duplicate var declarations in this method. Is probably better to declare them once at the start.

File: tests/symbols/spec.svg [hunk 7]
```diff
@@ -0,0 +1,16 @@
+<svg id="svgElement" width="500" height="500" xmlns="http://www.w3.org/2000/svg">
```
Comment (HackbrettXXX) [inline tests/symbols/spec.svg]: We should probably add more cases:
 * different transforms (on the `symbol`, on the `use`, "around" the `use`, ...)
 * style inheritance from "around" the `use` element

File: tests/symbols/spec.svg [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+  <symbol id="myDot" x="1" height="100" viewBox="0 0 80 100" >
```
Comment (HackbrettXXX) [inline tests/symbols/spec.svg]: 1px is really hard to see, make it e.g. 10px to really see an effect.

File: tests/symbols/spec.svg [hunk 9]
```diff
@@ -0,0 +1,16 @@
+<svg id="svgElement" width="500" height="500" xmlns="http://www.w3.org/2000/svg">
+  <symbol id="myDot" x="1" height="100" viewBox="0 0 80 100" >                                          
+    <circle cx="50" cy="50" r="50"/>
+  </symbol>
+  <symbol id="myDot-overflow" height="100" overflow="visible" viewBox="0 0 80 100" >                                          
+    <circle cx="50" cy="50" r="50"/>
+  </symbol>
+  
+  <use href="#myDot"   />
```
Comment (HackbrettXXX) [inline tests/symbols/spec.svg]: maybe you could align the circles a little better and make them distinguishable, so it's immediately clear which circle is which.

File: src/svg2pdf.js (hunk 1 above)
Comment (zenker981) [reply src/svg2pdf.js]: This is not my code... It already existed at v1.5.0 and I just moved it to an external function, but I will fix it there, sure ;)

Comment (HackbrettXXX) [reply src/svg2pdf.js]: The variable was declared earlier in the function scope:
```
var i, minX, minY, maxX, maxY, viewBox, vb, boundingBox;
```
In JavaScript `var` variables are not block-scoped, which means if a variable is declared anywhere within a function (e.g. nested in an if-block) it is available everywhere in that function.

Comment (zenker981) [reply src/svg2pdf.js]: Yup, it's fixed, I just wanted to explain that this was a copy/paste error (what doesn't make anything better, I know...)

**Pull request merged**